include/micm/util/cuda_dense_matrix.hpp

```diff
@@ -240,5 +236,18 @@ namespace micm
       }
       CHECK_CUDA_ERROR(micm::cuda::CopyToDeviceFromDevice(this->param_, other.param_), "cudaMemcpyDeviceToDevice");
     }
+
+    // Singleton to create the CUBLAS handle
+    static cublasHandle_t &GetCublasHandle(bool is_destructor)
+    {
+      static cublasHandle_t cublas_handle = NULL;
+      if (!cublas_handle && !is_destructor) CHECK_CUBLAS_ERROR(cublasCreate(&cublas_handle), "CUBLAS initialization failed...");
+      if (!cublas_handle && is_destructor)
+      {
+        CHECK_CUBLAS_ERROR(cublasDestroy(cublas_handle), "CUBLAS finalization failed...");
+        cublas_handle = NULL;
+      }
+      return cublas_handle;
+    }
```

Comment:
Would something like this work? It adds a custom deleter to the shared pointer so that we don't have to check the count of the shared pointer and we no longer need the boolean to decide when to destroy. Yes, I helped generate this with chatgpt


```
#include <memory>
#include <stdexcept>
#include <iostream>
#include <mutex>
#include <cublas_v2.h>

class CublasHandleSingleton {
public:
    // Get the cublas handle as a shared pointer with custom deleter
    static std::shared_ptr<cublasHandle_t> getHandle() {
        std::lock_guard<std::mutex> lock(getMutex()); // Lock the mutex
        static CublasHandleSingleton instance; // Ensures only one instance
        return instance.handle;
    }

private:
    std::shared_ptr<cublasHandle_t> handle;
    static std::mutex& getMutex() {
        static std::mutex mutex;
        return mutex;
    }

    // Private constructor to prevent instantiation
    CublasHandleSingleton() : handle(new cublasHandle_t, [](cublasHandle_t* handle) {
        if (*handle != nullptr) {
            cublasStatus_t status = cublasDestroy(*handle);
            if (status != CUBLAS_STATUS_SUCCESS) {
                std::cerr << "CUBLAS finalization failed: " << status << std::endl;
            }
            delete handle;
        }
    }) {
        cublasStatus_t status = cublasCreate(handle.get());
        if (status != CUBLAS_STATUS_SUCCESS) {
            throw std::runtime_error("CUBLAS initialization failed.");
        }
    }

    // Delete copy constructor and assignment operator
    CublasHandleSingleton(const CublasHandleSingleton&) = delete;
    CublasHandleSingleton& operator=(const CublasHandleSingleton&) = delete;

    // Destructor
    ~CublasHandleSingleton() = default; // Custom deleter in shared_ptr handles destruction
};

// Usage example
int main() {
    try {
        std::shared_ptr<cublasHandle_t> handle = CublasHandleSingleton::getHandle();
        // Use the handle for cuBLAS operations

        // When the shared_ptr goes out of scope, the custom deleter will be called
    } catch (const std::exception& e) {
        std::cerr << e.what() << std::endl;
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}

```

Reply:
@K20shores @mwaxmonsky thank you both for suggesting the implementation of singleton class for the cublas handle. After merging @mattldawson 's latest solver builder changes that address the GPU memory leak in the cuda rosenbrock test, the singleton class now works for the CUBLAS code and this PR is ready for your review again.